### PR TITLE
Enable post process with full precision using SubViewport

### LINKED ISSUE #54122: Vulkan: Cannot retrieve HDR texture from Viewport (LDR texture is returned instead)
### Godot version

4.0.dev

### System information

Vulkan API 1.2.162 - NVIDIA - GeForce GT 710

### Issue description

In Godot 3.x, when a Viewport was configured with `hdr` flag enabled, the result of `get_texture().get_data()` was an _RGBAHalf_ image.

In `master` branch, the `hdr` flag disappeared (because, as I understand, it's always HDR now), but calling `get_texture().get_image()` results in an _RGB8_ image, losing the high dynamic range.

### Steps to reproduce

* Create a SubViewport $vp
* In GDScript : `print($vp.get_texture().get_image().data.format)`

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by JonqsGames**
Fixes #54122

I was facing #54122 when trying to port a post effect i did in 3.4.4
This quick fix make the subviewport keep the `DATA_FORMAT_R16G16B16A16_SFLOAT` precision and disable tonemapping and linearization in the tonemapper shader.

The current implementation i did looks really bad but i'll try to clean it later.

- [x] Introduce a `keep_linear_3d` checkbox in SubViewport parameters
- [x] Pass the `keep_linear_3d` through SubViewports property  to the render target
- [x] Documentation (buffer format, disabled tonemapping...)

**Edit**:
I made a quick test project for this feature : 
[TestShaderPostProcess.zip](https://github.com/godotengine/godot/files/9020096/TestShaderPostProcess.zip)
Each cube is using the same shader that has a `float` uniform that is put into the color value.
Cube value of main scene are (-1.0, 0.0, 0.5, 1.0, 50.0).
Post process is simply putting a random color based on retrieved color value.
Base result (both `keep_linear` and `force_high_precision` disabled)
image
Result with both flag enabled (Vulkan):
image

### CODE REVIEW

**Comment by clayjohn** [comment]
A few things to keep in mind as your work to clean this up:
1. Tonemapping should not be disabled for SubViewports. SubViewports should have an option to keep 3D in linear space (in 3.x this option was called ``keep_3D_linear`` and was exposed as a checkbox on Viewports)
2. The setting should not be exposed as a parameter to ``render_scene`` it should be a property of SubViewports (internally a property of RenderingTarget). 
3. We should keep the default behaviour the same. Right now we keep 2D operations in sRGB space with a RGBA8 buffer to improve performance and memory usage. Users shouldnt pay the extra cost of a higher precision format unless it is actually necessary.

**Comment by Jojox** [comment]
Made a first pass of refactor on that. Remove the whole `is_sub_vp` behavior and added a keep_linear attribute instead.
Also refactored the way color format is choose for the render target, made the format available using setters and remove the hard coded part that was forcing the format.
I think the format could be simplified to one value (the actual buffer format) and all other format (srgb, image_format) could be choose based on it, this may be out of scope for this PR.

**Review by lyuma — COMMENTED** [review]
At some point, it might be good to make tonemapping entirely optional for the linear case...my assumption is you will pass this linear texture into another viewport at some point. if it runs tonemapping twice, that might look wrong. Furthermore, it would mean we can save a full-screen render pass on these linear targets which would improve performance.

There's a lot of plumbing here. My biggest concern is how these color_format and color_format_srgb are passed around everywhere and pollute the API, especially when they are often the same. It seems like much of the code is written under the assumption that those two formats are different. in the case of SFLOAT, from what I can tell, `_update_render_target` is still creating the alias for "color_format_srgb" despite it being the same format.

[ I'm talking about this code:
```cpp
              tex->rd_texture = RD::get_singleton()->texture_create_shared(view, rt->color);
                if (rt->color_format_srgb != RD::DATA_FORMAT_MAX) {
                        view.format_override = rt->color_format_srgb;
                        tex->rd_texture_srgb = RD::get_singleton()->texture_create_shared(view, rt->color);
                }
```
]
since color_format_srgb is not RD::DATA_FORMAT_MAX in the linear case. (and if it were, that might cause other bugs)

perhaps it would be better to move the linear flag here, and keep the current code otherwise. that way we wouldn't need those color format functions and all the duplicate code that picks both formats.

Overall, a lot of this change is plumbing. I'm not a rendering team lead, but the main problem I see with it is from making the color format code more confusing.

(there's also a question about do we really need another 128 bits of push constants for a single boolean? Well, there are already many other booleans being passed in this way, so I guess it's just an optimization job for another day.

**Comment by Jojox** [comment]
@lyuma For Tonemapping in this case i left the implementation as is. Linear is not doing anything.
Regarding the `color_format`, the change introduce here are needed because the user now has control over the buffer format in the editor (By the `keep_linear` checkbox). In my opinion user should have a total control on the `RenderTarget` buffer format for viewports but i think this belong to another PR and it needs to be thinked globaly. As @BastiaanOlij was saying, at the moment the whole viewport implementation asssume that it will be rendered in screen context so there is some shortcuts taken.
Exposing the `color_format` as any parameter in RenderTarget seems more logical to me and it also make the shortcuts more visible for the future refactoring.
I agree on the extra padding but i don't think it could be done another way. The `tonemap.glsl` shader is doing a lot more than tonemapping and several parameter could be factorize to different set of uniform and padding could be avoided but i think this belong to another PR.

File: servers/rendering/renderer_rd/effects/tone_mapper.h [hunk 1]
```diff
@@ -83,6 +83,8 @@ class ToneMapper {
 		float pixel_size[2]; //  8 - 104
 		uint32_t use_fxaa; //  4 - 108
 		uint32_t use_debanding; //  4 - 112
+		float pad[3]; //  12 - 114 # Padding to keep alignement
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/effects/tone_mapper.h:86]:
```suggestion
		float pad[3]; //  12 - 124
```

File: servers/rendering/renderer_viewport.cpp [hunk 2]
```diff
@@ -712,12 +712,36 @@ void RendererViewport::viewport_initialize(RID p_rid) {
 	Viewport *viewport = viewport_owner.get_or_null(p_rid);
 	viewport->self = p_rid;
 	viewport->render_target = RSG::texture_storage->render_target_create();
+	if (viewport->keep_linear) {
```
Comment (clayjohn) [inline servers/rendering/renderer_viewport.cpp:715]: Since this is a newly created viewport, ``keep_linear`` will always be false, so you don't need the conditional here, you can just initialize to default.

**Review by clayjohn — CHANGES_REQUESTED** [review]
Great work! You are getting much closer to a working implementation. 

Now is the time for us to solidify exactly what the API should look like as I can see you and I have differing ideas. 

My thinking was to just expose two booleans ``keep_3d_linear`` (2D is always sRGB), and ``force_high_precision``. 

``keep_3d_linear`` would be assigned to the render_target and would disable linear->sRGB in the tonemapper. You have this nearly implemented now and it should be fine. 

For ``force_high_precision`` my idea was that it would also get assigned to the render_target and then during allocation, the rendertarget would select between two formats based on what ``force_high_precision`` is set to. 

In my opinion, the above is about as simple as we can make this feature (it also worked similarly in 3.x).

However, I can see you are going in a slightly different direction, one which would provide the user with a little more flexibility by allowing them to override the default precision. I like this idea as well, and it aligns with our goal of giving users more control in 4.0. 

Accordingly, if you want to expose the ability to set the color format directly I would do it as follows:
1. Still have ``keep_3d_linear`` (this is needed in any case)
2. do not create ``force_high_precision`` instead create an ``override_color_format`` variable in Viewport which would call ``render_target_set_override_color_format`` (the same as your current ``render_target_set_color_format``). Then in ``update_render_target`` you check and see if ``override_color_format`` is used, if it is, then you set ``color_format``, ``color_format_srgb``, and ``image_format`` based on ``override_color_format``, if not, you just use the default. 
3. render_target would of course need a boolean ``use_override_color_format`` which is set to true if ``render_target_set_override_color_format`` is called with an argument that is not ``RD::DATA_FORMAT_R8G8B8A8_UNORM``

**Comment by Jojox** [comment]
Thanks for the detailed feedbacks. Both review converge for the `color_format` and i agree. I'll change that behavior to use `keep_precision` if it's fine for now to have two checkbox for that post process usecase. 
I choose to merge both behavior under the `keep_linear` because i thought the raw color doesn't make sense unless the buffer use full precision. Maybe i'm lacking some understanding on how the data is converted to RenderTarget expected color format.
To stay in scope of this PR, i'll limit my change to `keep_precision` toggle switching between the two scenario of color format. The idea of exposing can be done later if we want to give more control to the user.

File: drivers/png/png_driver_common.cpp [hunk 3]
```diff
@@ -125,6 +125,7 @@ Error png_to_image(const uint8_t *p_source, size_t p_size, bool p_force_linear,
 }
 
 Error image_to_png(const Ref<Image> &p_image, Vector<uint8_t> &p_buffer) {
+	ERR_FAIL_COND_V(p_image == NULL, FAILED);
```
Comment (clayjohn) [inline drivers/png/png_driver_common.cpp:128]: This should be removed and added to a separate pull request if it fixes a bug.

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 4]
```diff
@@ -1964,7 +1964,6 @@ RID RenderingDeviceVulkan::texture_create(const TextureFormat &p_format, const T
 	} else {
 		image_view_create_info.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
 	}
-
```
Comment (clayjohn) [inline drivers/vulkan/rendering_device_vulkan.cpp:1967]: This change should be reverted

File: doc/classes/RenderingServer.xml [hunk 5]
```diff
@@ -3161,6 +3161,14 @@
 				If [code]true[/code], rendering of a viewport's environment is disabled.
 			</description>
 		</method>
+		<method name="viewport_set_force_high_precision">
+			<return type="void" />
+			<argument index="0" name="viewport" type="RID" />
+			<argument index="1" name="enable" type="bool" />
+			<description>
+				If [code]true[/code], buffer format will be set to Half float.
```
Comment (clayjohn) [inline doc/classes/RenderingServer.xml:3169]:
```suggestion
				If [code]true[/code] and using the Vulkan-Clustered or Vulkan-mobile renderer, high precision will be used in the 2D rendering pipeline (3D is always high precision) and [method viewport_get_texture] will return an image with the [constant Image.FORMAT_RGBAH] format. Use with [method viewport_set_keep_linear] to obtain an accurate texture containing the rendering results from a frame. 
```

File: doc/classes/RenderingServer.xml [hunk 6]
```diff
@@ -3185,6 +3193,14 @@
 				Sets the viewport's global transformation matrix.
 			</description>
 		</method>
+		<method name="viewport_set_keep_linear">
+			<return type="void" />
+			<argument index="0" name="viewport" type="RID" />
+			<argument index="1" name="enable" type="bool" />
+			<description>
+				If [code]true[/code], skip delinearized pass in tonemapper shader.
```
Comment (clayjohn) [inline doc/classes/RenderingServer.xml:3201]:
```suggestion
				If [code]true[/code], skip linear to sRGB conversion during tonemapping. If used without [method viewport_set_force_high_precision] this can result in significant banding in the rendered output.
```

File: doc/classes/Viewport.xml [hunk 7]
```diff
@@ -196,6 +196,9 @@
 		<member name="disable_3d" type="bool" setter="set_disable_3d" getter="is_3d_disabled" default="false">
 			Disable 3D rendering (but keep 2D rendering).
 		</member>
+		<member name="force_high_precision" type="bool" setter="set_force_high_precision" getter="is_forced_high_precision" default="false">
+			If [code]true[/code], keep the buffer format for the viewport to Half Float
```
Comment (clayjohn) [inline doc/classes/Viewport.xml:200]: Adapt description from above here

File: doc/classes/Viewport.xml [hunk 8]
```diff
@@ -218,6 +221,10 @@
 		</member>
 		<member name="handle_input_locally" type="bool" setter="set_handle_input_locally" getter="is_handling_input_locally" default="true">
 		</member>
+		<member name="keep_linear" type="bool" setter="set_keep_linear" getter="is_keeping_linear" default="false">
+			If [code]true[/code], color will not be delinearized.
```
Comment (clayjohn) [inline doc/classes/Viewport.xml:225]: Adapt from description above

File: drivers/gles3/storage/texture_storage.cpp [hunk 9]
```diff
@@ -1324,6 +1324,14 @@ void TextureStorage::render_target_free(RID p_rid) {
 	}
 	render_target_owner.free(p_rid);
 }
+void TextureStorage::render_target_set_force_high_precision(RID p_render_target, bool p_force_high_precision) {
+	RenderTarget *rt = render_target_owner.get_or_null(p_render_target);
+	ERR_FAIL_COND(!rt);
+	if (rt->force_high_precision != p_force_high_precision) {
+		rt->force_high_precision = p_force_high_precision;
+		_update_render_target(rt);
+	}
```
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.cpp:1333]:
```suggestion
	rt->force_high_precision = p_force_high_precision;
```
There is no point in updating the render target if ``rt->force_high_precision`` is ignored anyway.

You may want to consider implementing ``force_high_precision`` and ``keep_linear`` in the GLES3 backend. It should not be much different than in the RendererRD backend

File: servers/rendering/renderer_scene_cull.cpp [hunk 10]
```diff
@@ -3226,7 +3226,6 @@ void RendererSceneCull::_render_scene(const RendererSceneRender::CameraData *p_c
 		occluders_tex = RSG::viewport->viewport_get_occluder_debug_texture(p_viewport);
 		prev_camera_data = RSG::viewport->viewport_get_prev_camera_data(p_viewport);
 	}
-
```
Comment (clayjohn) [inline servers/rendering/renderer_scene_cull.cpp:3229]: This should be reverted

**Review by clayjohn — CHANGES_REQUESTED** [review]
Looks mostly good! I left a few comments on style changes and documentation changes. But it is very close to being done. 

I think before merging, we should also have the settings working in the GLES3 renderer. It will work the exact same as in the RD renderer, so it should not be much more work. You can look at the GLES3 renderer in 3.x for the naming of the color format.

File: drivers/png/png_driver_common.cpp (hunk 3 above)
Comment (Jojox) [reply drivers/png/png_driver_common.cpp:128]: Yes i'm gonna move that to another PR. My editor crashed everytime i import an audio file, this fixed it but i'm not sure what the real bug is.

File: drivers/gles3/storage/texture_storage.cpp (hunk 9 above)
Comment (clayjohn) [reply drivers/gles3/storage/texture_storage.cpp:1333]: ``!spec->ok`` means that the engine is trying to draw with a shader that failed to compile. If you go to the top of your errors you should have a more complete error message that explains why the shader is failing to compile.

Comment (Jojox) [reply drivers/gles3/storage/texture_storage.cpp:1333]: Thanks i had indeed an error in scene shader. I'm unable to make my test project works in GLES3. It looks like something is clamping data between 0.0 and 1.0 outside of the linear call i found. I also noticed several issue in GLES3, for example my flag `keep_linear` when updated on my viewport is applied to all viewports (including the one used by editor).
Overall implementation of scene shaders in GLES3  is really tricky to  navigate in.
I'll get back to that later and create a simpler test project to share.

Comment (Jojox) [reply drivers/gles3/storage/texture_storage.cpp:1339]: By reading https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glTexImage2D.xhtml i understood that i had to use a custom format.
Example is now working in GLES3.
Not sure if i forgot some case in the `scene.glsl` shader but try a bunch of setup and everything works as expected.

**Comment by Jojox** [comment]
I did the requested changes. The only thing left is the behavior in edior for GLES3. I think this is due to other issue not related to this PR.

**Comment by fire** [comment]
I also don't think gles3 is a blocker for this pr.

**Comment by lyuma** [comment]
I might be inclined to agree that this needn't wait for GLES3. My guess is that there is not any work needed for GLES3. Note that GLES3 is a low-end renderer (much like GLES2 was in Godot 3.x), so it likely uses gamma rendering in non-HDR
(in Godot shader terms, I would expect `OUTPUT_IS_SRGB` to be true in GLES3 when using Godot 4.x).

It might be good to have an explicit half- or full-precision float render target mode for GLES3 (useful for calculations / simulation viewports), but that should be proposed and then added as a separate PR if there is demand, I think.

**Comment by Jojox** [comment]
Actually the feature works in GLES3 with the same behavior than in Vulkan. I just noticed that the behavior in editor is not the same. When changing the flags, the viewport does not update properly sometimes. I also had the case where the `keep_linear` flag was applied to all viewport/windows in the editor. But when reloading the scene or running it everything behave as expected.

File: drivers/gles3/storage/texture_storage.cpp [hunk 11]
```diff
@@ -1130,11 +1130,17 @@ void TextureStorage::_update_render_target(RenderTarget *rt) {
 		rt->fbo = system_fbo;
 		return;
 	}
-
-	rt->color_internal_format = rt->is_transparent ? GL_RGBA8 : GL_RGB10_A2;
-	rt->color_format = GL_RGBA;
-	rt->color_type = rt->is_transparent ? GL_BYTE : GL_UNSIGNED_INT_2_10_10_10_REV;
-	rt->image_format = Image::FORMAT_RGBA8;
+	if (rt->force_high_precision) {
+		rt->color_internal_format = rt->is_transparent ? GL_RGBA16F : GL_RGB16F;
```
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.cpp:1134]: Mobile devices don't typically support RGBA16 or RGB16 (some do through extensions though). So we can't rely on support being available. We will have to add a config setting ``framebuffer_half_float_supported`` and then fall back to low precision if it is false. You can initialize the config the same way we do in 3.x, e.g. in storage/config.cpp:
```
#ifdef GLES_OVER_GL
	config.framebuffer_float_supported = true;
#else
	framebuffer_float_supported = config.extensions.has("GL_EXT_color_buffer_float");
#endif

Comment (JonqsGames) [reply drivers/gles3/storage/texture_storage.cpp:1200]: Had to do a big rebase, everything looks fine but hopefully i didn't missed something :crossed_fingers:
Would'nt it be better to just use the `#ifdef` to ignore the force_high_precision ? Since the high precision for thoses devices is actually the default behavior.

Comment (JonqsGames) [reply drivers/gles3/storage/texture_storage.cpp:1201]: Added a check for the support of float framebuffer

File: drivers/gles3/storage/texture_storage.cpp [hunk 12]
```diff
@@ -1196,11 +1197,17 @@ void TextureStorage::_update_render_target(RenderTarget *rt) {
...
 	}
-
-	rt->color_internal_format = rt->is_transparent ? GL_RGBA8 : GL_RGB10_A2;
-	rt->color_format = GL_RGBA;
-	rt->color_type = rt->is_transparent ? GL_BYTE : GL_UNSIGNED_INT_2_10_10_10_REV;
-	rt->image_format = Image::FORMAT_RGBA8;
+	if (rt->force_high_precision && config->framebuffer_float_supported) {
+		rt->color_internal_format = rt->is_transparent ? GL_RGBA16F : GL_RGB16F;
+		rt->color_format = GL_RGBA;
+		rt->color_type = GL_BYTE;
```
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.cpp:1203]: Type should be ``GL_HALF_FLOAT`` and I am now just realizing I missed something else. We need to check for half float support as well not just float support as the [GL_EXT_color_buffer_float](https://registry.khronos.org/OpenGL/extensions/EXT/EXT_color_buffer_float.txt) extension only makes RGBA16F a renderable format, it does not include RGB16F, for that you need [GL_EXT_color_buffer_half_float](https://registry.khronos.org/OpenGL/extensions/EXT/EXT_color_buffer_half_float.txt). So ``framebuffer_float_supported`` should indicate that both GL_EXT_color_buffer_float and GL_EXT_color_buffer_half_float are available.

**Review by clayjohn — CHANGES_REQUESTED** [review]
My apologies, I missed a few things on my last review. After this change it should be good to go!

File: drivers/gles3/storage/texture_storage.cpp (hunk 12 above)
Comment (JonqsGames) [reply drivers/gles3/storage/texture_storage.cpp]: Made the fix. I didn't knew that `RGB16F` format was considered so exotic. I could try to find a way to avoid it in this case but i'm not sure how yet.
Thanks again for the carefull feedbacks !

**Review by clayjohn — DISMISSED** [review]
I'm happy with this implementation. Would be nice for @BastiaanOlij and @reduz to weigh in before merging as this is a big change.

**Comment by BastiaanOlij** [comment]
As for making it work as we did in Godot 3 this seems sound, but I've never really found we're on the right track with this to begin with. Possibly something to discuss in more detail in a render meeting when we have more time.

The core issue here are the assumptions that:
A) we're always combining 2D and 3D
B) we're always outputting to screen
C) we're always outputting color data

None of these assumptions are true and how we render our content differs greatly. Even 2D output can be valid input as a texture used in 3D where colors are not sRGB or where we even don't want RGBA8 as our texture format. Think of rendering noise or height map data. 

Now in Vulkan we do have more control over how we treat color data both when writing out and when reading but I think we need far more control over configuring the data type of our (sub)viewport and what sort of rendering we're doing.

This comes back down to my issue with having `RenderTarget` and `RenderBuffers` as 2D and 3D constructs within the render server and believing this is the wrong approach. Yes if a viewport is used for final output and we're doing both 3D and 2D rendering, we need to render to a high precision 3D buffer first, then blit and tonemap into our SRGBA8 buffer, then render 2D, as we do now. But we have other variations possible too especially when we're producing an intermediate result.

Anyway, enough ranting :)

**Comment by akien-mga** [comment]
As a side note, the commits will need to be squashed before merging (see [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase)).

**Comment by reduz** [comment]
I don't think this is the right way to tackle this problem to be honest.
* After tonemap, whathever it happens can' t be considered post processing.
* I agree that it should be able to postprocess _before_ tonemap, but this is not the way.
* I think we need a way to have viewports in HDR, but setting them to linear is not the way and a hack. Instead, what needs to be done is add proper HDR support for 2D.

I think this is stuff that we need to discuss properly and that will most likely be done for 4.1, the time to add this sort of features has run out for 4.0.

**Comment by JonqsGames** [comment]
Totally agree on the fact that this a hack. I still think that it's better than nothing. The fact that this hack was (kind of) available in 3.x version may be an issue for some people.
Giving global control over the buffer formats and default processing steps for viewports would be better but require a lot of changes and a global plan to keep things coherents.
On the state of viewport now, it's handy to have them but it can be misleading for new user to understand the many restriction around them. In my opinion along with some documentation this hack can unlock some usage on post processing that will help plan the next steps.

**Comment by clayjohn** [comment]
@JonqsGames reduz and I discussed his view further in the rendering channel in [Rocketchat](chat.godotengine.org). The main problems with the current approach (that you and I decided on) is that it conflicts with eventually adding proper HDR support for HDR monitors. For that, we need to add a viewport option that tonemaps 3D into the proper color space and then renders 2D in the proper color space (most likely linear with a tonemapping operator at the end). This current design would end up having to be removed leading to breaking changes.

Instead what reduz suggests (and I now agree) is the best approach is to expose the 3D rendering buffer through the viewport API so users can select whether they want their final texture to be the 2D render target, or the 3D render buffer. This has the added benefit that the entire 2D rendering pipeline can be skipped if the user just wants the 3D render buffer (the tonemap shader won't even run).

We already do something similar for XR with the ``use_xr`` property. Reduz suggests that we expand that into an enum:
```
enum ViewportMode {
    2D_AND_3D,
    3D,
    XR
};
```

**Comment by clayjohn** [comment]
Superseded by https://github.com/godotengine/godot/pull/70970
